src/sinks/opendal_common.rs

```diff
@@ -0,0 +1,269 @@
+use futures::{stream::BoxStream, StreamExt};
+use opendal::Operator;
+use snafu::Snafu;
+use tower::Service;
+use tracing::Instrument;
+use vector_common::{
+    finalization::{EventStatus, Finalizable},
+    request_metadata::{MetaDescriptive, RequestMetadata},
+};
+use vector_core::{
+    internal_event::CountByteSize,
+    sink::StreamSink,
+    stream::{BatcherSettings, DriverResponse},
+    ByteSizeOf,
+};
+
+use crate::{
+    codecs::{Encoder, Transformer},
+    event::{Event, EventFinalizers},
+    internal_events::SinkRequestBuildError,
+    sinks::{
+        util::{
+            metadata::RequestMetadataBuilder, partitioner::KeyPartitioner,
+            request_builder::EncodeResult, Compression, RequestBuilder, SinkBuilderExt,
+        },
+        BoxFuture,
+    },
+};
+
+pub struct OpenDALSink {
+    op: Operator,
+    request_builder: OpenDALRequestBuilder,
+    partitioner: KeyPartitioner,
+    batcher_settings: BatcherSettings,
+}
+
+impl OpenDALSink {
+    pub fn new(
+        op: Operator,
+        request_builder: OpenDALRequestBuilder,
+        partitioner: KeyPartitioner,
+        batcher_settings: BatcherSettings,
+    ) -> Self {
+        Self {
+            op,
+            request_builder,
+            partitioner,
+            batcher_settings,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl StreamSink<Event> for OpenDALSink {
+    async fn run(
+        self: Box<Self>,
+        input: futures_util::stream::BoxStream<'_, Event>,
+    ) -> Result<(), ()> {
+        self.run_inner(input).await
+    }
+}
+
+impl OpenDALSink {
+    async fn run_inner(self: Box<Self>, input: BoxStream<'_, Event>) -> Result<(), ()> {
+        let partitioner = self.partitioner;
+        let settings = self.batcher_settings;
+
+        let builder_limit = NonZeroUsize::new(64);
+        let request_builder = self.request_builder;
+
+        input
+            .batched_partitioned(partitioner, settings)
+            .filter_map(|(key, batch)| async move {
+                // We don't need to emit an error here if the event is dropped since this will occur if the template
+                // couldn't be rendered during the partitioning. A `TemplateRenderingError` is already emitted when
+                // that occurs.
+                key.map(move |k| (k, batch))
+            })
+            .request_builder(builder_limit, request_builder)
+            .filter_map(|request| async move {
+                match request {
+                    Err(error) => {
+                        emit!(SinkRequestBuildError { error });
+                        None
+                    }
+                    Ok(req) => Some(req),
+                }
+            })
+            .into_driver(OpenDALService::new(self.op.clone()))
+            // TODO: set protocl with services scheme instead hardcoded file
+            .protocol("file")
+            .run()
+            .await
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct OpenDALService {
```

Comment:
Really sorry, it has been pointed out to me that according to the (naming standards](https://rust-lang.github.io/api-guidelines/naming.html), this should actually be `OpenDal`:

> In UpperCamelCase, acronyms and contractions of compound words count as one word: use Uuid rather than UUID, Usize rather than USize or Stdin rather than StdIn